src/kubernetes/state/instrumenting.rs

```diff
@@ -0,0 +1,355 @@
+//! An instrumenting state wrapper.
+
+use crate::internal_events::kubernetes::instrumenting_state as internal_events;
+use async_trait::async_trait;
+use futures::future::BoxFuture;
+
+/// A [`super::Write`] implementatiom that wraps another [`super::Write`] and
+/// adds instrumentation.
+pub struct Writer<T> {
+    inner: T,
+}
+
+impl<T> Writer<T> {
+    /// Take a [`super::Write`] and return it wrapped with [`Self`].
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+}
+
+#[async_trait]
+impl<T> super::Write for Writer<T>
+where
+    T: super::Write + Send,
+{
+    type Item = <T as super::Write>::Item;
+
+    async fn add(&mut self, item: Self::Item) {
+        emit!(internal_events::StateItemAdded);
```

Comment:
It is possible, however it decouples the concerns. Moving this logic into the reflector would mean that we have to test it there, and apply unneeded complexity.
Also, having this separate allows us to choose at which point of the storages stack do we want to emit the events. Thus, it won't be possible to emit on effective state updates if we move this logic to reflector - only on the ones perceived by the reflector. This makes a difference when we have delayed deletes.

Reply:
It makes sense, and the ability to choose were we would emit events is useful, but 350+ lines for this minor feature that only we would be using, is too much. Our algorithm + tests for graph validation of topology has the same number of lines.

Reply:
I'm afraid that by moving this logic somewhere else we won't be reducing the loc, but rather increasing it.

The majority (~300 lines) are tests, that are currently ignored. The reason they're ignored is that we don't have a proper way to test internal events atm. We can just yank them for now, but eventually, I want to test internal events emission.

Moving the logic elsewhere means that the tests will also have to go elsewhere. Adding another test case at reflector will alone cost more than 200 LOCs, and it won't be as atomically unit-focused.

Also, technically we'd want them to be at the state layer - after the delayed deletion, so we count effective state updates (as those will correspond to the observed behavior), rather than received events. This is a small but very important detail.

As you can see, it's not that simple.

I agree that the file is big though. In fact, this is something that bothered me too while I was working on the tests. I needed to advance, so I decided to postpone figuring out a more elegant way to organize the tests. Maybe it's time to give it a shot. Or yank the tests. But I'd prefer that either of those to moving the logic elsewhere - in my view it'll do more harm than good.

Reply:
I suppose this comes down to how we should handle testing of internal events, personally I'm for minimal amount, in this case none, but that's a whole different issue so let's stop here.

Reply:
I agree. I appreciate going the extra mile with tests. In general I'm a big advocate of testing, but they have a cost. And in this case, I do not think we need to test internal event emission unless it is core the functioning of the feature. For example, if we were actually relying on those events in some way within Vector.